File: TestClient_cpp/WorkThread.cpp
#include "StdAfx.h"

#include <MinerConnection.h>
#include <Hash.h>

#include "WorkThread.h"

void ScryptHashes(WorkThread* tinfo)
{
    ScryptData* scryptData = ScryptInit(tinfo->mWork);

    tinfo->mSolutionFound = false;
    unsigned int hashesDone = 0;
    __int64 end = tinfo->mHashStart + tinfo->mHashCount;
    for( __int64 i = tinfo->mHashStart; i < end; i++ )
    {
        if( WaitForSingleObject(tinfo->mKill, 0) == WAIT_OBJECT_0 )
            break;

        hashesDone++;
        bool success = ScryptHash(i, tinfo->mWork, scryptData);
        if( success )
        {
            tinfo->mSolution = i;
            tinfo->mSolutionFound = true;
            break;
        }
    }

    delete scryptData;

    tinfo->mHashesDone = hashesDone;
}

DWORD WINAPI HashWorkThread(void* param)
{
    WorkThread* tinfo = (WorkThread*)param;

    while( 1 )
    {
        if( WaitForSingleObject(tinfo->mKill, 0) == WAIT_OBJECT_0 )
            break;

        if( WaitForSingleObject(tinfo->mWorkToDo, 0) == WAIT_OBJECT_0 )
        {
            // There is work to be done!
            //printf("Thread(%d) starting work\n", tinfo->mThreadId);
            switch( tinfo->mWork->currency )
            {
                case 0:     // SHA256^2
                    break;
                case 1:     // Scrypt
                case 2:                    
                    ScryptHashes(tinfo);
                    break;
                default:  
                    break;
            }

            // All done
            int success = SetEvent(tinfo->mWorkDone);
            //printf("Thread(%d) done with work\n", tinfo->mThreadId);
        }

        Sleep(1);
    }

    SetEvent(tinfo->mDead);
    return 0;
}


WorkThread::WorkThread(void)
{
    mWork = new Work();
    mHashCount = 0;

    mKill = CreateEvent(NULL, FALSE, FALSE, NULL);
    mDead = CreateEvent(NULL, FALSE, FALSE, NULL);
    mWorkToDo = CreateEvent(NULL, FALSE, FALSE, NULL);
    mWorkDone = CreateEvent(NULL, TRUE, FALSE, NULL);
    mThreadHandle = CreateThread(NULL, 0, HashWorkThread, this, 0, &mThreadId);
}

WorkThread::~WorkThread(void)
{
}

void WorkThread::StartWork(unsigned int hashStart, unsigned int hashCount, Work* work)
{  
    mWork[0] = *work;
   
    mHashStart = hashStart;
    mHashCount = hashCount;

    mHashesDone = 0;
    mSolutionFound = false;

    ResetEvent(mWorkDone);
    SetEvent(mWorkToDo);
}

bool WorkThread::WorkDone()
{
    return (WaitForSingleObject(mWorkDone, 0) == WAIT_OBJECT_0);
}

File: MinerLib_cs/WorkThread.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace MinerLib_cs
{
    public class WorkThread
    {
        Work mWork;
        uint mHashStart;
        uint mHashCount;
        uint mHashesDone;
        uint mSolution;
        bool mSolutionFound;

        Thread mThread;
        bool mWorkToDo;
        bool mWorkDone;

        public WorkThread()
        {
            mWork = new Work();
            mHashCount = 0;

            mThread = new Thread(new ThreadStart(WorkFunction));
            mThread.Start();
        }

        public void Destroy()
        {
            mThread.Abort();
        }

        void WorkFunction()
        {
            while (true)
            {
                if (mWorkToDo)
                {
                    // There is work to be done!
                    switch (mWork.hashAlgorithm)
                    {
                        case Work.Algorithm.SHA256:     // SHA256^2
                            break;
                        case Work.Algorithm.Scrypt:
                            ScryptHashes();
                            break;
                        default:
                            break;
                    }

                    // All done
                    mWorkDone = true;
                }
                                
                Thread.Sleep(1);
            }
        }

        public void StartWork(uint hashStart, uint hashCount, Work work)
        {
            mHashStart = hashStart;
            mHashCount = hashCount;
            mWork.Copy(work);

            mHashesDone = 0;
            mSolutionFound = false;
            
            mWorkDone = false;
            mWorkToDo = true;
        }

        public bool IsWorkDone()
        {
            return mWorkDone;
        }

        public uint HashesDone
        {
            get { return mHashesDone; }
        }

        public uint Solution
        {
            get { return mSolution; }
        }

        public bool SolutionFound
        {
            get { return mSolutionFound; }
        }

        void ScryptHashes()
        {
            ulong end = mHashStart + mHashCount;
            Scrypt scryptData = new Scrypt(mWork);
            for( ulong i = mHashStart; i < end; i++ )
            {
                mHashesDone++;
                bool success = scryptData.Hash((uint)i);
                if( success )
                {
                    mSolution = (uint)i;
                    mSolutionFound = true;
                    break;
                }
            }
        }
    }
}
